src/include/simeng/config/SimInfo.hh

```diff
@@ -0,0 +1,245 @@
+  void extractValues() {
+    // Get ISA type and set the corresponding architectural fileset
+    std::string isa;
+    validatedConfig_["Core"]["ISA"] >> isa;
+    if (isa == "AArch64") {
+      isa_ = config::ISA::AArch64;
+      // Define system registers
+      sysRegisterEnums_ = {arm64_sysreg::ARM64_SYSREG_DCZID_EL0,
+                           arm64_sysreg::ARM64_SYSREG_FPCR,
+                           arm64_sysreg::ARM64_SYSREG_FPSR,
+                           arm64_sysreg::ARM64_SYSREG_TPIDR_EL0,
+                           arm64_sysreg::ARM64_SYSREG_MIDR_EL1,
+                           arm64_sysreg::ARM64_SYSREG_CNTVCT_EL0,
+                           arm64_sysreg::ARM64_SYSREG_PMCCNTR_EL0,
+                           arm64_sysreg::ARM64_SYSREG_SVCR};
+    } else if (isa == "rv64") {
+      isa_ = config::ISA::RV64;
+      // Define system registers
+      sysRegisterEnums_ = {};
+    }
+
+    // Initialise architectural reg structures by using the reset
+    // functionality
+    resetArchRegStruct();
+
+    // Get Simulation mode
+    std::string mode;
+    validatedConfig_["Core"]["Simulation-Mode"] >> mode;
+    if (mode == "emulation") {
+      mode_ = config::simMode::emulation;
+      modeStr_ = "Emulation";
+    } else if (mode == "inorderpipelined") {
+      mode_ = config::simMode::inorder;
+      modeStr_ = "In-Order Pipelined";
+    } else if (mode == "outoforder") {
+      mode_ = config::simMode::outoforder;
+      modeStr_ = "Out-of-Order";
+    }
+
+    // Get if the special files directory should be created
+    validatedConfig_["CPU-Info"]["Generate-Special-Dir"] >> genSpecialFiles_;
+  }
+
+  /** Function used to reset the architectural register file structure. */
+  void resetArchRegStruct() {
+    // Given some register quantities rely on Config file arguments (SME
+    // relies on SVL), it is possible that if the config was to change the
+    // register quantities would be incorrect. This function provides a way to
+    // reset the Architectural register structure.
+    if (isa_ == config::ISA::AArch64) {
+      uint16_t numSysRegs = static_cast<uint16_t>(sysRegisterEnums_.size());
+      // Set the size of SME ZA in bytes by dividing the SVL by 8
+      uint16_t ZAbits;
+      validatedConfig_["Core"]["Streaming-Vector-Length"] >> ZAbits;
+      const uint16_t ZAsize = ZAbits / 8;
+      archRegStruct_ = {
+          {8, 32},          // General purpose
+          {256, 32},        // Vector
+          {32, 17},         // Predicate
+          {1, 1},           // NZCV
+          {8, numSysRegs},  // System
+          {256, ZAsize},    // Matrix (Each row is a register)
+      };
+    } else if (isa_ == config::ISA::RV64) {
+      uint16_t numSysRegs = static_cast<uint16_t>(sysRegisterEnums_.size());
+      archRegStruct_ = {
+          {8, 32},          // General purpose
+          {8, 32},          // Floating Point
+          {8, numSysRegs},  // System
+      };
```

Comment:
After spending some time with this issue, I'm not seeing it as you are. I'm not seeing how the maintenance nor difficulty of adding new ISAs comes into play here. It just serves to fragment the configuration of a model. The points I do understand are following the directory structure layout and the separation of arch and µarch.  However, the additional complexity of a solution that works here I don't think is worth it for the sake of fragmenting the source of model configuration options.

Offline solutions such as having header files in the arch folders to store ISA-specific information have been explored. However, due to config information only being available at runtime it partially invalidated their use as you have to deal with ISA-specific information within SimInfo anyway (which is then accessed later through SimInfo) so you might as well keep it local to SimInfo. This then begs the question of why separate it at all.

I'm not opposed to the idea but can't personally find a viable/agreeable solution. If you have any additional solutions, happy to explore them.

Reply:
I think a way to do it is to have a templated function called `genArchRegStruct` . It will have template specialisation for each of the ISAs we support and we can then define the body of each specialisation in the relevant Architecture.cc file. Roughly this would be done as:

##### SimInfo.hh
```
class SimInfo {

......

template <class A>
std::vector<simeng::RegisterFileStructure>  genArchRegStruct();

// We now need to move the definition of resetArchRegStruct from Siminfo.hh to SimInfo.cc, so the linker
// can resolve the dependency properly
void resetArchRegStruct();

......

}
```
##### src/lib/arch/riscv/Architecture.cc
```
....

#include "SimInfo.hh"
template <>
std::vector<simeng::RegisterFileStructure> SimInfo::genArchRegStruct<arch::riscv::Architecture>() {
     uint16_t numSysRegs = static_cast<uint16_t>(sysRegisterEnums_.size());
      return {
          {8, 32},          // General purpose
          {8, 32},          // Floating Point
          {8, numSysRegs},  // System
      };
};

...
```
##### src/lib/arch/aarch64/Architecture.cc
```
// Similar definition as in src/lib/arch/riscv/Architecture.cc
```
##### src/lib/SimInfo.cc (New File)
```
#include "SimInfo"

void SimInfo::resetArchRegStruct() {
   if (isa_ == config::ISA::AArch64) {
      archRegStruct_ = genArchRegStruct<arch::aarch64::Architecture>();
    } else if (isa_ == config::ISA::RV64) {
      archRegStruct_ = genArchRegStruct<arch::riscv::Architecture>();
}

```

This way we "move" the definition to the relevant Architecture.cc file. Specifics may vary given SimEng has a complex file structure and you **MAY** have to resolve some circular dependencies, but the idea remains the same.

Rough working example [here](https://replit.com/join/lstcdcegbx-rahatmuneeb).

Thoughts @dANW34V3R @FinnWilkinson

Reply:
Same can be done for other functions ISA-specific variables like `sysRegisterEnums`

Reply:
I still think this doesn't address the problems raised. It's very complex for the functionality it provides. You still have to deal with non-agnostic data from the config during runtime (e.g. SVL value) as well.

Reply:
I am not sure what you mean by (Sorry if I've misunderstood):
> You still have to deal with non-agnostic data from the config during **runtime**

These template method implementations are instance methods of the SimInfo class and they will be able to access member variables in the SimInfo class. (ISA specific or agnostic). Which parts of the problem are still not addressed (Again, sorry if I've misunderstood the problem)?

As mentioned in my comment the specifics may vary, and my suggestion was only there to show a possible approach to solving the problems mentioned in @dANW34V3R  's comment without creating any extra mediating classes:

> **I am not a fan of having this all defined outside of the arch folders (inc. system register enums) as it makes it hard to maintain and difficult for anyone adding a new ISA**. Although I can see the problem here as ZAsize is dependent on the config. A not particularly thought through idea is having some sort of static architectural info class which is queried just to keep all of the arch-specific info under the arch folders. This class could contain other info e.g. register file structure, system register structure, which register is the stack pointer, which registers are used for syscall arguments etc.

It does add a little complexity but also it helps readability and provides separation of concern.

```
void SimInfo::resetArchRegStruct() {
   if (isa_ == config::ISA::AArch64) {
      archRegStruct_ = genArchRegStruct<arch::aarch64::Architecture>();
    } else if (isa_ == config::ISA::RV64) {
      archRegStruct_ = genArchRegStruct<arch::riscv::Architecture>();
}
```

Ultimately the decision rests with you, I am happy with whatever you decide.

Reply:
I'd argue that from a readability perspective, the solution doesn't provide anything different from the current implementation. For the "You still have to deal with non-agnostic data from the config during runtime", with this solution you'd still have to parse the SVL config option (depending on the selected ISA) so why not just parse the rest of the ISA-specific data in the same place (as opposed to fragmenting it across multiple files).

Reply:
I agree with the point you're making, but you can parse all ISA-specific configs through a single function following the same structure I've stated above (and still provide separation of concern). 
As far as this is concerned:
>  so why not just parse the rest of the ISA-specific data in the same place (as opposed to fragmenting it across multiple files.

If you feel that parsing ISA-specific options in an ISA-agnostic class (SimInfo) won't lead to maintenance-related problems in the future especially when we add more ISAs (because ISA-related changes will now have to be reflected in three places SimInfo.hh and Architecture.cc/hh); then keep it as is. 

As mentioned I'm happy to keep it as is, because this isn't even a problem. However, I thought I'd give my perspective given I do agree with the points Dan made.